### PR TITLE
[Plugin] Remote Clock based off Telemetry

### LINKED ISSUE #3983: Implement new "Remote Clock"
Right now Open MCT supports a ["local clock"](https://github.com/nasa/openmct/blob/master/src/plugins/utcTimeSystem/LocalClock.js) out of the box for realtime telemetry operations. This clock ticks on a local time source (ie. [a recurring `setTimeout` which just pulls the current time  from `Date.now()`](https://github.com/nasa/openmct/blob/master/src/plugins/utcTimeSystem/LocalClock.js#L69)).

This approach only supports so called "wall clock" time, and does not support simulations or telemetry playback of historical or future time periods.

To support these use cases we need a new clock that ticks in real-time, just like the local clock, but which gets a timing signal from a remote source instead of using `Date.now()`.

The new remote clock plugin should be configurable with a telemetry point from which to receive the time data.

eg.
``` JavaScript
openmct.install(new openmct.plugins.RemoteClock({namespace: 'something', key:'telemetry-key'});
```
On activation the remote clock will
1. Use telemetry metadata to select the appropriate timestamp on the configured telemetry point to fetch time data from (it will be the `domain` field that has the same `key` as the active time system.
2. Request the latest time using a historical query (ie. `openmct.telemetry.request`)
3. Subscribe to the configured telemetry point in order to receive timing signals on an ongoing basis.

When a new telemetry datum is received the Remote Clock will ["tick" with the latest time (as provided by the server) as ms since 01/01/1970](https://github.com/nasa/openmct/blob/master/src/plugins/utcTimeSystem/LocalClock.js#L70) (ie JavaScript time).

Most of the logic in the RemoteClock will be the same as logic in the [LocalClock](https://github.com/nasa/openmct/blob/master/src/plugins/utcTimeSystem/LocalClock.js#L70). The only difference is that instead of a `setTimeout` the tick will be triggered by the telemetry subscription callback. You might consider creating a shared parent class that they both extend.

### PR DESCRIPTION
**Pull request by jvigliotta**
RemoteClock plugin. Adds a plugin that accepts a domainObject identifier that the clock will pull time from for "ticks".

closes #3983

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?

### Author Checklist

* [x] Changes address original issue?
* [x] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

### CODE REVIEW

File: src/plugins/remoteClock/RemoteClock.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            strategy: 'latest'
+        }).then(data => {
+            this._processDatum(data[data.length - 1]);
+        });
+    }
+
+    _processDatum(datum) {
+        let time = this.parseTime(datum);
+
+        if (time > this.lastTick) {
```
Comment (akhenry) [inline src/plugins/remoteClock/RemoteClock.js:94]: beautiful. \**chef's kiss*\*

File: src/plugins/remoteClock/RemoteClockSpec.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * Open MCT Web includes source code licensed under additional open source
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+define(["./RemoteClock"], function (RemoteClock) {
```
Comment (akhenry) [inline src/plugins/remoteClock/RemoteClockSpec.js:23]: Let's use ES6 imports, not requirejs

File: src/plugins/remoteClock/RemoteClockSpec.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        // let telemetryObject = {
+        //     name: 'Telemetry Object',
+        //     identifier: {
+        //         namespace: 'telemetry',
+        //         key: 'object'
+        //     }
+        // };
+
+        beforeEach(function () {
+            clock = new RemoteClock();
```
Comment (akhenry) [inline src/plugins/remoteClock/RemoteClockSpec.js:35]: We should test the plugin itself, not the `RemoteClock` directly. Testing the plugin checks that it is installing the clock ok and that it is available to use from the time API.

**Review by akhenry — CHANGES_REQUESTED** [review]
Looks awesome. A couple of comments inline, all really minor things. I do think we need a couple more tests though (we should be testing the plugin, not the RemoteClock class).

File: src/plugins/remoteClock/RemoteClock.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+        this.lastTick = 0;
+
+        this._processDatum = this._processDatum.bind(this);
+    }
+
+    start() {
+        this.openmct.time.on('timeSystem', this._timeSystemChange);
+        this.openmct.objects.get(this.identifier).then((domainObject) => {
+            this.remoteTelemetryObject = domainObject;
```
[the remark this answers is not in the document]
Comment (jvigliotta) [reply src/plugins/remoteClock/RemoteClock.js:56]: I like it... I love it... I want... to let you know that, that is what I changed it to.

File: src/plugins/remoteClock/RemoteClockSpec.js (hunk 3 above)
Comment (jvigliotta) [reply src/plugins/remoteClock/RemoteClockSpec.js:35]: Working on that now!

Comment (jvigliotta) [reply src/plugins/remoteClock/RemoteClockSpec.js:35]: I don't know why I yelled that.

File: src/plugins/remoteClock/RemoteClock.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            this._timeSystemChange();
+            this._requestLatest();
+            this._subscribe();
+        }).catch((error) => {
+            throw new Error(error);
+        });
+    }
+
+    stop() {
+        this.openmct.time.off('timeSystem', this._timeSystemChange);
```
Comment (akhenry) [inline src/plugins/remoteClock/RemoteClock.js:69]: Good catch! ⚾

File: src/plugins/remoteClock/RemoteClockSpec.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        it('will request metadata and set up formatters', () => {
+            expect(remoteClock.metadata).toEqual(metadata);
+            expect(metadata.value).toHaveBeenCalled();
+            expect(openmct.telemetry.getValueFormatter).toHaveBeenCalledWith(metadataValue);
+        });
+
+        it('will request the latest datum for the object it received and process the datum returned', () => {
+            expect(openmct.telemetry.request).toHaveBeenCalledWith(remoteClock.timeTelemetryObject, REQ_OPTIONS);
+            expect(remoteClock._processDatum).toHaveBeenCalledWith(reqDatum);
```
Comment (akhenry) [inline src/plugins/remoteClock/RemoteClockSpec.js:137]: I tend to try and avoid creating dependencies on private functions from test specs, and instead just test external behavior. That way if the details of the class change but its behavior doesn't, the spec does not get invalidated.

I would listen to `openmct.time.on('bounds', ...)` and make sure it is called with the expected bounds (and `isTick = true`)

File: src/plugins/remoteClock/RemoteClockSpec.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            expect(remoteClock.tick).toHaveBeenCalledWith(reqDatum.key);
+        });
+
+        it('will set up subscriptions correction', () => {
+            expect(remoteClock._unsubscribe).toBeDefined();
+            expect(openmct.telemetry.subscribe).toHaveBeenCalledWith(remoteClock.timeTelemetryObject, remoteClock._processDatum);
```
Comment (akhenry) [inline src/plugins/remoteClock/RemoteClockSpec.js:143]: 👨‍🍳 🤏 💋

**Review by akhenry — CHANGES_REQUESTED** [review]
Just one tiny change on one of the tests. We should try and test external behavior whenever possible and not code dependencies on internal details of our classes.

File: src/plugins/remoteClock/RemoteClockSpec.js (hunk 6 above)
Comment (jvigliotta) [reply src/plugins/remoteClock/RemoteClockSpec.js:137]: Gotcha, makes sense, I'll update that!

**Review by akhenry — APPROVED** [review]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Has associated issue been labelled `unverified`? (only applicable if this PR closes the issue) Y

**Pull request merged**